# Avocado Datenanalyse

![Getting Started](./avocado.webp)

## KPI (Key Performance Indicator)

Es handelt sich um ein Leistungskennzahl, die verwendet wird, um den Erfolg oder Fortschritt in Bezug auf ein spezifisches Ziel zu messen. KPIs sind besonders in Bereichen wie Wirtschaft, Management, Marketing und Datenanalyse wichtig, um die Leistung verschiedener Prozesse, Strategien oder Aktivitäten zu bewerten.

**Eigenschaften eines guten KPI sind:**
- Spezifisch: Der KPI sollte klar und präzise definieren, was gemessen wird.
- Messbar: Der KPI sollte quantifizierbar sein, d.h. er sollte in Zahlen oder Prozentsätzen ausgedrückt werden können.
- Erreichbar: Das Ziel des KPI sollte realistisch und erreichbar sein.
- Relevant: Der KPI sollte in direktem Zusammenhang mit den Zielen des Unternehmens oder Projekts stehen.
- Zeitgebunden: Der KPI sollte an eine bestimmte Zeitspanne gebunden sein, z.B. monatlich, quartalsweise, jährlich.

Angenommen, du bist ein Avocado-Händler und möchtest wissen, wie erfolgreich dein Verkauf in einer bestimmten Region ist. Dann könnten KPIs wie zum Beispiel "Durchschnittlicher Verkaufspreis" entscheidend sein, um den Markttrend und die Nachfrage zu bewerten. Eine signifikante Änderung in diesen KPIs könnte darauf hinweisen, dass Anpassungen an der Preisstrategie oder Marketingkampagnen notwendig sind.
</br>
</br>

**Welche KPIs könnten für uns interessant sein?**
- Durchschnittspreis (Average Price): Der durchschnittliche Verkaufspreis von Avocados in einer bestimmten Region oder in einem bestimmten Zeitraum. Dieser KPI hilft, die Preisentwicklung zu überwachen.
- Gesamtvolumen (Total Volume): Die Gesamtmenge der verkauften Avocados. Dies kann anzeigen, wie gut sich das Produkt in verschiedenen Regionen verkauft und wie stark die Nachfrage ist.
- Verkaufsvolumen nach Typ (Volume by Type): Vergleich des Verkaufsvolumens zwischen konventionellen und biologischen Avocados. Dieser KPI hilft, den Marktanteil und die Beliebtheit der beiden Typen zu verstehen.
- Preis-Volumen-Korrelation: Untersuchung der Beziehung zwischen Preis und Verkaufsvolumen. Ein negativer Zusammenhang kann darauf hindeuten, dass höhere Preise zu geringeren Verkaufszahlen führen.
- Jährliche Wachstumsrate: Die prozentuale Veränderung des Durchschnittspreises oder Verkaufsvolumens im Vergleich zum Vorjahr. Dieser KPI zeigt das Wachstum oder den Rückgang des Marktes über die Jahre.
</br>
</br>

**Warum sind KPIs wichtig?**
- Zielverfolgung: KPIs ermöglichen es, Fortschritte in Richtung festgelegter Ziele zu messen.
- Leistungsbewertung: Sie helfen dabei, die Leistung verschiedener Geschäftsbereiche, Produkte oder Strategien zu bewerten.
- Entscheidungsfindung: KPIs liefern wertvolle Einblicke, die bei der Entscheidungsfindung und strategischen Planung unterstützen.
- Frühwarnsystem: KPIs können frühzeitig auf Probleme hinweisen, sodass rechtzeitig Maßnahmen ergriffen werden können.


Zusammenfassend sind KPIs essentielle Werkzeuge zur Messung und Verbesserung der Leistung in einem Geschäfts- oder Analysekontext. Sie bieten klare, messbare Ziele und helfen, Entscheidungen auf fundierten Daten zu basieren.


**Mögliche KPIs für die Avocado-Datenanalyse**

1. Durchschnittlicher Verkaufspreis (Average Price):
   1. Beschreibung: Durchschnittlicher Preis pro Avocado (konventionell und biologisch).
   2. Fragestellung: Wie entwickeln sich die durchschnittlichen Preise über die Jahre und in verschiedenen Regionen?
   3. Ziel: Preisentwicklung und Preisstabilität überwachen, um Markttrends zu verstehen.
2. Gesamtverkaufsvolumen (Total Volume):
   1. Beschreibung: Die gesamte Anzahl verkaufter Avocados in einer bestimmten Region oder einem bestimmten Zeitraum.
   2. Fragestellung: Wie viel Volumen wird in den verschiedenen Regionen verkauft? Gibt es Spitzen in bestimmten Jahreszeiten?
   3. Ziel: Nachfrage und Saisonalität identifizieren.
3. Verkaufsvolumen nach Typ (Volume by Type):
   1. Beschreibung: Vergleich des Verkaufsvolumens zwischen konventionellen und biologischen Avocados.
   2. Fragestellung: Welche Art von Avocados ist beliebter (konventionell oder biologisch) und wie variiert dies in verschiedenen Regionen?
   3. Ziel: Marktanteile der verschiedenen Produktkategorien bewerten.
4. Umsatz pro Region (Revenue per Region):
   1. Beschreibung: Gesamtumsatz in einer bestimmten Region (Durchschnittspreis * Gesamtvolumen).
   2. Fragestellung: Welche Regionen generieren den höchsten Umsatz?
   3. Ziel: Regionen mit hoher Verkaufsleistung identifizieren und Marketingmaßnahmen gezielt einsetzen.
5. Jährliche Wachstumsrate (Annual Growth Rate):
   1. Beschreibung: Prozentuale Änderung des Verkaufsvolumens und des Preises im Vergleich zum Vorjahr.
   2. Fragestellung: Gibt es ein Wachstum oder einen Rückgang im Verkauf und Preis? Welche Regionen wachsen am schnellsten?
   3. Ziel: Langfristige Wachstums- und Rückgangstrends überwachen.
6. Preis-Volumen-Korrelation (Price-Volume Correlation):
   1. Beschreibung: Beziehung zwischen dem Preis und dem Verkaufsvolumen.
   2. Fragestellung: Gibt es eine negative Korrelation, die darauf hinweist, dass höhere Preise zu geringeren Verkaufszahlen führen?
   3. Ziel: Verständnis des Einflusses von Preisänderungen auf die Verkaufszahlen.
7. Saisonale Nachfrage (Seasonal Demand):
   1. Beschreibung: Veränderung des Verkaufsvolumens und Preises über das Jahr hinweg.
   2. Fragestellung: Gibt es saisonale Spitzen, z.B. zur Weihnachtszeit oder im Sommer?
   3. Ziel: Bessere Planung der Lagerbestände und Marketingaktionen basierend auf saisonaler Nachfrage.
8. Marktanteil (Market Share):
   1. Beschreibung: Anteil des Verkaufsvolumens einer Region im Vergleich zum Gesamtvolumen.
   2. Fragestellung: Welche Region hat den größten Marktanteil? Gibt es Veränderungen im Marktanteil über die Jahre?
   3. Ziel: Wettbewerbsfähigkeit in verschiedenen Regionen bewerten.

**Mögliche Fragestellungen für eine tiefere Analyse:**
1. Preisentwicklung: Wie hat sich der Preis für Avocados in den letzten Jahren entwickelt? Gibt es Unterschiede in der Preisentwicklung zwischen konventionellen und biologischen Avocados?
2. Regionale Unterschiede: Welche Regionen zeigen die höchste Nachfrage nach Avocados? Welche Regionen haben die höchsten Preise?
3. Saisonale Trends: Gibt es bestimmte Jahreszeiten, in denen der Avocado-Verkauf signifikant ansteigt oder abnimmt?
4. Verhältnis von Preis und Volumen: Wie beeinflusst eine Preisänderung das Verkaufsvolumen? Gibt es einen optimalen Preis, bei dem das Volumen maximiert wird?
5. Nachfrage nach Bioprodukten: Wie hat sich die Nachfrage nach biologischen Avocados im Vergleich zu konventionellen entwickelt? Welche Faktoren könnten diesen Unterschied erklären?
6. Umsatzentwicklung pro Jahr: Wie entwickelt sich der Umsatz insgesamt und in den einzelnen Regionen? Welche Regionen sind für das größte Wachstum verantwortlich?


## 1. Datenimport und Übersicht

Zunächst importieren wir die Daten und verschaffen uns einen Überblick über die Struktur der Daten.

In [18]:
import pandas as pd

# Datei laden
FILE_PATH = "avocado.csv" # Pfad als Konstante (Best Practise)
avocado_data = pd.read_csv(FILE_PATH)

# Erste Zeilen anzeigen
avocado_data.head()

,Unnamed: 0,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,0.0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015.0,Albany
1,1.0,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015.0,Albany
2,2.0,2015-12-13,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015.0,Albany
3,3.0,2015-12-06,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015.0,Albany
4,4.0,2015-11-29,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015.0,Albany


PLU-Nummern (Price Look-Up Codes) sind vier- oder fünfstellige Codes, die auf Obst und Gemüse verwendet werden, um sie an der Kasse korrekt zu identifizieren. Diese Codes helfen dabei, verschiedene Sorten, Größen oder Anbaumethoden (konventionell vs. biologisch) eines Produkts zu unterscheiden. Sie sind weltweit standardisiert und werden hauptsächlich im Einzelhandel verwendet.

1. `PLU 4046`:
   1. Beschreibung: Diese Nummer bezieht sich auf kleine Hass-Avocados.
2. `PLU 4225`:
   1. Beschreibung: Diese Nummer steht für mittelgroße Hass-Avocados.
3. `PLU 4770`:
   1. Beschreibung: Diese Nummer wird für große Hass-Avocados verwendet.

## 2. Grundlegende Dateninformationen

Im nächsten Schritt untersuchen wir die Struktur der Daten, um Datentypen, Anzahl der Einträge und fehlende Werte zu identifizieren.
- `info()`: Zeigt Informationen über Datentypen und nicht-null Werte an.
- `describe()`: Liefert eine statistische Übersicht, einschließlich Mittelwert, Minimum und Maximum für numerische Spalten.

In [19]:
# Struktur und Datenarten anzeigen
avocado_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18253 entries, 0 to 18252
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    18251 non-null  float64
 1   Date          18252 non-null  object 
 2   AveragePrice  18251 non-null  object 
 3   Total Volume  18250 non-null  float64
 4   4046          18250 non-null  float64
 5   4225          18250 non-null  float64
 6   4770          18250 non-null  float64
 7   Total Bags    18250 non-null  float64
 8   Small Bags    18250 non-null  float64
 9   Large Bags    18250 non-null  float64
 10  XLarge Bags   18250 non-null  float64
 11  type          18249 non-null  object 
 12  year          18249 non-null  float64
 13  region        18249 non-null  object 
dtypes: float64(10), object(4)
memory usage: 1.9+ MB


**Interpretation:**

- `<class 'pandas.core.frame.DataFrame'>`: Der Datensatz ist ein pandas DataFrame.
- `RangeIndex: 18253 entries, 0 to 18252`: Der DataFrame enthält 18.253 Einträge (Zeilen), beginnend von Index 0 bis 18.252.
- `Data columns (total 14 columns)`: Der DataFrame besteht aus insgesamt 14 Spalten.
- `Non-Null Count`: Die Anzahl der nicht-null Werte in jeder Spalte. Diese Werte zeigen an, wie viele Einträge in der jeweiligen Spalte vorhanden sind und keine fehlenden Werte (NaN) aufweisen.
- `Dtype`: Der Datentyp der jeweiligen Spalte.
- `memory usage: 1.9+ MB`: Der DataFrame benötigt etwa 1,9 MB Speicherplatz im Arbeitsspeicher.
- `float64 (10 Spalten)`: Repräsentiert numerische Werte mit Dezimalstellen, typischerweise für Preise und Verkaufsvolumen.
- `object (4 Spalten)`: Repräsentiert Zeichenfolgen, typischerweise für Texte wie Datumsangaben oder Kategorisierungen (z.B. Typ, Region).

**Interpretation der Spalteninformationen:**

0. `Unnamed: 0`
   1. Nicht-null Werte: 18.251 von 18.253.
   2. Datentyp: float64.
   3. Interpretation: Diese Spalte enthält möglicherweise einen zusätzlichen Index oder fortlaufende Zahlen. Es fehlen jedoch 2 Werte, was auf Unstimmigkeiten oder Fehler hinweisen könnte.
1. `Date`
   1. Nicht-null Werte: 18.252 von 18.253.
   2. Datentyp: object.
   3. Interpretation: Diese Spalte enthält Datumseinträge. Der Datentyp object deutet darauf hin, dass die Daten als Zeichenfolgen vorliegen und in ein Datetime-Format umgewandelt werden sollten. Ein Eintrag fehlt.
2. AveragePrice
   1. Nicht-null Werte: 18.251 von 18.253.
   2. Datentyp: object.
   3. Interpretation: Diese Spalte sollte die durchschnittlichen Verkaufspreise enthalten. Der Datentyp object zeigt an, dass einige Einträge möglicherweise nicht numerisch sind, was zu Problemen bei Berechnungen führen kann. 2 Werte fehlen.
3. `Total Volume`
   1. Nicht-null Werte: 18.250 von 18.253.
   2. Datentyp: float64.
   3. Interpretation: Diese Spalte enthält das Gesamtvolumen der verkauften Avocados. 3 Einträge fehlen.
4. `4046`, `4225`, `4770`
   1. Nicht-null Werte: 18.250 von 18.253.
   2. Datentyp: float64.
   3. Interpretation: Diese Spalten repräsentieren spezifische Avocado-Typen oder -Kategorien basierend auf PLU-Codes (Produktnummern). Sie geben das Verkaufsvolumen für jede Kategorie an. 3 Werte fehlen jeweils.
5. `Total Bags`, `Small Bags`, `Large Bags`, `XLarge Bags`
   1. Nicht-null Werte: 18.250 von 18.253.
   2. Datentyp: float64.
   3. Interpretation: Diese Spalten enthalten Informationen über die Anzahl der verkauften Avocados in verschiedenen Verpackungsgrößen. Es fehlen jeweils 3 Werte.
6. `type`
   1. Nicht-null Werte: 18.249 von 18.253.
   2. Datentyp: object.
   3. Interpretation: Diese Spalte gibt den Typ der Avocados an, entweder "conventional" (konventionell) oder "organic" (biologisch). 4 Werte fehlen.
7. `year`
   1. Nicht-null Werte: 18.249 von 18.253.
   2. Datentyp: float64.
   3. Interpretation: Diese Spalte enthält das Jahr der Verkaufsdaten. 4 Werte fehlen.
8. `region`
   1. Nicht-null Werte: 18.249 von 18.253.
   2. Datentyp: object.
   3. Interpretation: Diese Spalte gibt die geografische Region an, in der die Daten gesammelt wurden. Auch hier fehlen 4 Werte.

In [20]:
# Statistische Zusammenfassung der numerischen Daten
avocado_data.describe()

,Unnamed: 0,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,year
count,18251.00000,1.825000e+04,1.825000e+04,1.825000e+04,1.825000e+04,1.825000e+04,1.825000e+04,1.825000e+04,18250.000000,18249.000000
mean,24.22985,8.505982e+05,2.929924e+05,2.951384e+05,2.283848e+04,2.396269e+05,1.821853e+05,5.433535e+04,3106.256292,2016.147899
std,15.48189,3.453456e+06,1.264956e+06,1.204089e+06,1.074613e+05,9.862168e+05,7.461592e+05,2.439596e+05,17692.424825,0.939938
min,0.00000,8.456000e+01,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000e+00,0.000000,2015.000000
25%,10.00000,1.083963e+04,8.535750e+02,3.008097e+03,0.000000e+00,5.089083e+03,2.850323e+03,1.275800e+02,0.000000,2015.000000
50%,24.00000,1.073655e+05,8.643200e+03,2.905888e+04,1.849750e+02,3.974118e+04,2.635161e+04,2.648225e+03,0.000000,2016.000000
75%,38.00000,4.329527e+05,1.110087e+05,1.501663e+05,6.242055e+03,1.107811e+05,8.333621e+04,2.201828e+04,132.432500,2017.000000
max,52.00000,6.250565e+07,2.274362e+07,2.047057e+07,2.546439e+06,1.937313e+07,1.338459e+07,5.719097e+06,551693.650000,2018.000000


- `count`: Anzahl der nicht fehlenden Werte in jeder Spalte.
- `mean`: Der Durchschnittswert jeder Spalte.
- `std`: Die Standardabweichung, die die Streuung der Werte zeigt.
- `min`: Der minimale Wert in jeder Spalte.
- `25%`, `50%`, `75%` (Quartile): Die Werte an den entsprechenden Quartilen, die die Verteilung beschreiben.
- `max`: Der maximale Wert in jeder Spalte.

**Interpretation der einzelnen Spalten:**

1. `Total Volume`:
   1. std: Eine hohe Standardabweichung von 3.453.456 deutet darauf hin, dass das Verkaufsvolumen stark variiert.
   2. min: Das minimale Verkaufsvolumen ist 84,56, was auf sehr geringe Verkaufszahlen in einigen Fällen hinweist.
   3. max: Das maximale Verkaufsvolumen beträgt 62.505.650, was auf ein sehr hohes Verkaufsvolumen in bestimmten Fällen hinweist.
2. `4046`, `4225`, `4770`:
   1. min: Der minimale Wert ist 0, was bedeutet, dass es Einträge gibt, bei denen kein Volumen für diese Kategorien verzeichnet wurde.
   2. max: Es gibt hohe Maximalwerte, insbesondere für PLU 4046 und 4225, was auf eine hohe Nachfrage hinweist.
3. `Total Bags`, `Small Bags`, `Large Bags`, `XLarge Bags`:
   1. mean: Die meisten verkauften Taschen sind „Small Bags“ mit einem Durchschnitt von 182.185, gefolgt von „Large Bags“ (54.335) und „XLarge Bags“ (3.106).
   2. min: Der minimale Wert ist 0, was darauf hinweist, dass es Fälle gibt, in denen keine Avocados in Taschen verkauft wurden.
   3. max: Der maximale Wert zeigt, dass „Small Bags“ (13.384.590) am meisten verwendet werden, gefolgt von „Large Bags“ und „XLarge Bags“.
4. `year`:
   1. mean: Das durchschnittliche Jahr ist etwa 2016, was bedeutet, dass die Daten über mehrere Jahre gesammelt wurden.
   2. min und max: Die Daten reichen von 2015 bis 2018, was die Zeitspanne der Datenerfassung zeigt.

**Wichtige Erkenntnisse:**

1. Die hohe Standardabweichung in Total Volume zeigt, dass das Verkaufsvolumen stark variiert. Dies deutet auf sehr unterschiedliche Verkaufszahlen zwischen den Regionen oder Zeiträumen hin.
2. PLU 4046 und 4225 haben deutlich höhere Maximalwerte im Vergleich zu PLU 4770. Das könnte auf eine höhere Nachfrage oder Beliebtheit dieser spezifischen Avocadosorten hindeuten.
3. Die meisten Avocados werden in „Small Bags“ verkauft, was möglicherweise auf eine bevorzugte Verpackungsgröße im Einzelhandel hinweist.
4. Die Daten umfassen die Jahre 2015 bis 2018. Die mittlere Verteilung nahe 2016 zeigt, dass es eine relativ gleichmäßige Datenerfassung über die Jahre gibt.
5. Die minimalen Werte von 0 in den PLU- und Taschenkategorien deuten darauf hin, dass es Datensätze gibt, in denen keine Verkäufe für bestimmte Kategorien verzeichnet wurden. Diese sollten überprüft werden, um sicherzustellen, dass die Daten vollständig sind.


In [21]:
# Fehlende Werte überprüfen
avocado_data.isnull().sum()

Unnamed: 0      2
Date            1
AveragePrice    2
Total Volume    3
4046            3
4225            3
4770            3
Total Bags      3
Small Bags      3
Large Bags      3
XLarge Bags     3
type            4
year            4
region          4
dtype: int64

## 3. Bereinigung der Daten

Hier werden die Daten bereinigt, indem unnötige Spalten entfernt und fehlerhafte Werte korrigiert werden.

### 3.1 Auf Duplikate überprüfen

- Die Methode `duplicated()` prüft jede Zeile im DataFrame avocado_data darauf, ob sie ein Duplikat ist. Es gibt eine Serie von True oder False Werten zurück, wobei True bedeutet, dass die Zeile ein Duplikat ist.
- Die Methode `sum()` summiert die True-Werte der duplicated()-Methode.

In [22]:
# Boolsche Maske erzeugen
avocado_data.duplicated()

0        False
1        False
2        False
3        False
4        False
         ...  
18248    False
18249    False
18250    False
18251    False
18252    False
Length: 18253, dtype: bool

In [23]:
# Anzahl der Duplikate ausgeben
avocado_data.duplicated().sum() # Wenn "0" dann keine Duplikate vorhanden

np.int64(0)

### 3.2 Entfernen überflüssiger Spalten

Die Spalte `Unnamed: 0` schein nicht sinnvoll zu sein.

In [24]:
# Entfernen der überflüssigen Spalte 'Unnamed: 0'
avocado_data = avocado_data.drop(['Unnamed: 0'], axis=1) # axis=1: Gibt an, dass eine Spalte entfernt werden soll. axis=0 würde stattdessen eine Zeile entfernen.
avocado_data

,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015.0,Albany
1,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015.0,Albany
2,2015-12-13,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015.0,Albany
3,2015-12-06,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015.0,Albany
4,2015-11-29,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015.0,Albany
...,...,...,...,...,...,...,...,...,...,...,...,...,...
18248,2018-02-04,1.63,17074.83,2046.96,1529.20,0.00,13498.67,13066.82,431.85,0.0,organic,2018.0,WestTexNewMexico
18249,2018-01-28,1.71,13888.04,1191.70,3431.50,0.00,9264.84,8940.04,324.80,0.0,organic,2018.0,WestTexNewMexico
18250,2018-01-21,1.87,13766.76,1191.92,2452.79,727.94,9394.11,9351.80,42.31,0.0,organic,2018.0,WestTexNewMexico
18251,2018-01-14,1.93,16205.22,1527.63,2981.04,727.01,10969.54,10919.54,50.00,0.0,organic,2018.0,WestTexNewMexico


### 3.3 Konvertieren von timestamps in den richtigen Datentyp

Das Konvertieren von Spalten mit Zeitangaben in das Datetime-Format ist ein wichtiger Schritt, um sicherzustellen, dass die Daten in der Date-Spalte korrekt als Datum interpretiert werden.
Sobald die Date-Spalte in ein datetime-Format konvertiert wurde, können spezifische Zeitreihenanalysen durchgeführt werden. Man kann Trends über Tage, Monate oder Jahre hinweg erkennen und analysieren, wie sich beispielsweise der Durchschnittspreis von Avocados im Laufe der Zeit verändert hat.

In [25]:
# Konvertieren der 'Date'-Spalte in das Datetime-Format
avocado_data['Date'] = pd.to_datetime(avocado_data['Date'])

ValueError: time data "2018" doesn't match format "%Y-%m-%d", at position 169. You might want to try:
    - passing `format` if your strings have a consistent format;
    - passing `format='ISO8601'` if your strings are all ISO8601 but not necessarily in exactly the same format;
    - passing `format='mixed'`, and the format will be inferred for each element individually. You might want to use `dayfirst` alongside this.


Der Fehler `ValueError: time data "2018" doesn't match format "%Y-%m-%d"` tritt auf, wenn die Date-Spalte gemischte Formate enthält, die nicht mit dem erwarteten Format übereinstimmen. In diesem Fall ist das Problem, dass einige Einträge in der Date-Spalte nur das Jahr (z.B. "2018") und keine vollständigen Datumsangaben enthalten.
</br>
</br>
Überprüfe, welche unterschiedlichen Formate in der Date-Spalte vorhanden sind:

In [26]:
print(avocado_data['Date'].unique())  # Zeigt die einzigartigen Datumswerte an

['2015-12-27' '2015-12-20' '2015-12-13' '2015-12-06' '2015-11-29'
 '2015-11-22' '2015-11-15' '2015-11-08' '2015-11-01' '2015-10-25'
 '2015-10-18' '2015-10-11' '2015-10-04' '2015-09-27' '2015-09-20'
 '2015-09-13' '2015-09-06' '2015-08-30' '2015-08-23' '2015-08-16'
 '2015-08-09' '2015-08-02' '2015-07-26' '2015-07-19' '2015-07-12'
 '2015-07-05' '2015-06-28' '2015-06-21' '2015-06-14' '2015-06-07'
 '2015-05-31' '2015-05-24' '2015-05-17' '2015-05-10' '2015-05-03'
 '2015-04-26' '2015-04-19' '2015-04-12' '2015-04-05' '2015-03-29'
 '2015-03-22' '2015-03-15' '2015-03-08' '2015-03-01' '2015-02-22'
 '2015-02-15' '2015-02-08' '2015-02-01' '2015-01-25' '2015-01-18'
 '2015-01-11' '2015-01-04' '2016-12-25' '2016-12-18' '2016-12-11'
 '2016-12-04' '2016-11-27' '2016-11-20' '2016-11-13' '2016-11-06'
 '2016-10-30' '2016-10-23' '2016-10-16' '2016-10-09' '2016-10-02'
 '2016-09-25' '2016-09-18' '2016-09-11' '2016-09-04' '2016-08-28'
 '2016-08-21' '2016-08-14' '2016-08-07' '2016-07-31' '2016-07-24'
 '2016-07-

Konvertiere die Date-Spalte mit `pd.to_datetime()` unter Verwendung von `errors='coerce'`, um fehlerhafte Einträge in `NaT` (Not a Time) zu konvertieren:

In [27]:
# Dies hilft, die ungültigen Datumswerte in NaT umzuwandeln, sodass sie im weiteren Verlauf einfach identifiziert und behandelt werden können.
avocado_data['Date'] = pd.to_datetime(avocado_data['Date'], errors='coerce')
avocado_data['Date'].unique()  # Zeigt die einzigartigen Datumswerte an

<DatetimeArray>
['2015-12-27 00:00:00', '2015-12-20 00:00:00', '2015-12-13 00:00:00',
 '2015-12-06 00:00:00', '2015-11-29 00:00:00', '2015-11-22 00:00:00',
 '2015-11-15 00:00:00', '2015-11-08 00:00:00', '2015-11-01 00:00:00',
 '2015-10-25 00:00:00',
 ...
 '2018-03-04 00:00:00', '2018-02-25 00:00:00', '2018-02-18 00:00:00',
 '2018-02-11 00:00:00', '2018-02-04 00:00:00', '2018-01-28 00:00:00',
 '2018-01-21 00:00:00', '2018-01-14 00:00:00', '2018-01-07 00:00:00',
                 'NaT']
Length: 170, dtype: datetime64[ns]

Nach der Umwandlung können ungültige Datumswerte einfach identifiziert werden:

In [28]:
invalid_dates = avocado_data[avocado_data['Date'].isna()]
invalid_dates


,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
18206,NaT,Tampa,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18207,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
18208,NaT,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Diese Werte können entweder korrigiert oder aus dem DataFrame entfernt werden:

In [29]:
# Entfernen aller Zeilen, die ungültige Werte (NaN/NaT) in der 'Date'-Spalte enthalten
avocado_data = avocado_data.dropna(subset=['Date'])

# Überprüfen, ob die ungültigen Zeilen entfernt wurden
avocado_data[avocado_data['Date'].isna()]

,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region


In [30]:
# Für den Überblick
avocado_data

,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region
0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015.0,Albany
1,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015.0,Albany
2,2015-12-13,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015.0,Albany
3,2015-12-06,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015.0,Albany
4,2015-11-29,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015.0,Albany
...,...,...,...,...,...,...,...,...,...,...,...,...,...
18248,2018-02-04,1.63,17074.83,2046.96,1529.20,0.00,13498.67,13066.82,431.85,0.0,organic,2018.0,WestTexNewMexico
18249,2018-01-28,1.71,13888.04,1191.70,3431.50,0.00,9264.84,8940.04,324.80,0.0,organic,2018.0,WestTexNewMexico
18250,2018-01-21,1.87,13766.76,1191.92,2452.79,727.94,9394.11,9351.80,42.31,0.0,organic,2018.0,WestTexNewMexico
18251,2018-01-14,1.93,16205.22,1527.63,2981.04,727.01,10969.54,10919.54,50.00,0.0,organic,2018.0,WestTexNewMexico


In [31]:
# Nochmal die Informationen prüfen
avocado_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18250 entries, 0 to 18252
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          18250 non-null  datetime64[ns]
 1   AveragePrice  18250 non-null  object        
 2   Total Volume  18250 non-null  float64       
 3   4046          18250 non-null  float64       
 4   4225          18250 non-null  float64       
 5   4770          18250 non-null  float64       
 6   Total Bags    18250 non-null  float64       
 7   Small Bags    18250 non-null  float64       
 8   Large Bags    18250 non-null  float64       
 9   XLarge Bags   18250 non-null  float64       
 10  type          18249 non-null  object        
 11  year          18249 non-null  float64       
 12  region        18249 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(3)
memory usage: 1.9+ MB


### 3.4 Entfernen von Einträgen mit fehlenden Werten

In [32]:
avocado_data = avocado_data.dropna()
# Nochmal die Informationen prüfen
avocado_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18249 entries, 0 to 18252
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          18249 non-null  datetime64[ns]
 1   AveragePrice  18249 non-null  object        
 2   Total Volume  18249 non-null  float64       
 3   4046          18249 non-null  float64       
 4   4225          18249 non-null  float64       
 5   4770          18249 non-null  float64       
 6   Total Bags    18249 non-null  float64       
 7   Small Bags    18249 non-null  float64       
 8   Large Bags    18249 non-null  float64       
 9   XLarge Bags   18249 non-null  float64       
 10  type          18249 non-null  object        
 11  year          18249 non-null  float64       
 12  region        18249 non-null  object        
dtypes: datetime64[ns](1), float64(9), object(3)
memory usage: 1.9+ MB


### 3.5 Konvertieren von strings in numberische Werte

In [33]:
avocado_data['AveragePrice'] = pd.to_numeric(avocado_data['AveragePrice'], errors='coerce')
# Nochmal die Informationen prüfen
avocado_data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 18249 entries, 0 to 18252
Data columns (total 13 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          18249 non-null  datetime64[ns]
 1   AveragePrice  18249 non-null  float64       
 2   Total Volume  18249 non-null  float64       
 3   4046          18249 non-null  float64       
 4   4225          18249 non-null  float64       
 5   4770          18249 non-null  float64       
 6   Total Bags    18249 non-null  float64       
 7   Small Bags    18249 non-null  float64       
 8   Large Bags    18249 non-null  float64       
 9   XLarge Bags   18249 non-null  float64       
 10  type          18249 non-null  object        
 11  year          18249 non-null  float64       
 12  region        18249 non-null  object        
dtypes: datetime64[ns](1), float64(10), object(2)
memory usage: 1.9+ MB


C:\Users\olexa\AppData\Local\Temp\ipykernel_15556\521992367.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  avocado_data['AveragePrice'] = pd.to_numeric(avocado_data['AveragePrice'], errors='coerce')


## 4. KPIs und detaillierte Analysen

Jetzt folgen die Analysen mit den gewünschten KPIs und den Methoden groupby und pivot table, wie bereits oben beschrieben.

### 4.1 Durchschnittspreis (Average Price) nach Jahr und Region:

In [34]:
average_price_by_year = avocado_data.groupby('year')['AveragePrice'].mean()
average_price_by_year

year
2015.0    1.375590
2016.0    1.338640
2017.0    1.515128
2018.0    1.347531
Name: AveragePrice, dtype: float64

### 4.2 Gesamtvolumen (Total Volume) nach Jahr und Region:

In [35]:
total_volume_pivot = avocado_data.pivot_table(values='Total Volume', index='region', columns='year', aggfunc='sum')
total_volume_pivot

year,2015.0,2016.0,2017.0,2018.0
region,,,,
Albany,4.029896e+06,5.264336e+06,5.231582e+06,1.541986e+06
Atlanta,2.323170e+07,2.832688e+07,2.881512e+07,8.231422e+06
BaltimoreWashington,4.064558e+07,4.089380e+07,4.101563e+07,1.215890e+07
Boise,3.784357e+06,4.653510e+06,4.760561e+06,1.214760e+06
Boston,2.745499e+07,3.057132e+07,3.061067e+07,8.637006e+06
BuffaloRochester,5.893235e+06,7.205112e+06,7.740306e+06,2.123818e+06
California,3.014074e+08,3.261884e+08,3.176445e+08,8.374131e+07
Charlotte,9.487343e+06,1.058672e+07,1.189559e+07,3.585890e+06
Chicago,4.149472e+07,3.961264e+07,4.098075e+07,1.161423e+07


**Interpretation:**

- Regionen wie California und TotalUS haben deutlich höhere Verkaufsvolumen als andere Regionen. Dies zeigt, dass in diesen Regionen entweder eine hohe Nachfrage oder ein großes Angebot herrscht.
- Kleinere Regionen wie Boise oder Spokane haben im Vergleich dazu viel geringere Verkaufsvolumen.
- In vielen Regionen ist eine signifikante Veränderung des Verkaufsvolumens von Jahr zu Jahr zu erkennen. Zum Beispiel:
  - In LosAngeles stieg das Volumen von 148.220.700 Einheiten im Jahr 2015 auf 158.481.200 Einheiten im Jahr 2017, bevor es 2018 auf 38.052.380 Einheiten sank. Dies könnte auf Marktveränderungen, Angebotsprobleme oder eine Änderung der Nachfrage hinweisen.
  - In TotalUS stieg das Verkaufsvolumen kontinuierlich von 2015 bis 2017, bevor es 2018 einen deutlichen Rückgang verzeichnete.
- Besondere Beobachtungen:
  - Einige Regionen wie California, Texas und der Südosten (Southeast) verzeichnen besonders hohe Verkaufsvolumen, was auf eine starke Nachfrage oder ein hohes Angebot in diesen Gebieten hinweist.
  - Die Werte für TotalUS geben eine gute Übersicht über den gesamten Avocado-Verkauf in den USA und können als Referenzpunkt für das gesamte Marktvolumen verwendet werden.

### 4.3 Verkaufsvolumen nach Typ (Volume by Type):

In [36]:
volume_by_type_pivot = avocado_data.pivot_table(values='Total Volume', index='year', columns='type', aggfunc='sum')
volume_by_type_pivot

type,conventional,organic
year,,
2015.0,4.296599e+09,8.886943e+07
2016.0,4.690250e+09,1.306401e+08
2017.0,4.766166e+09,1.681399e+08
2018.0,1.334206e+09,4.853227e+07


**Interpretation:**

- Starkes Wachstum des biologischen Segments:
  - Von 2015 bis 2017 zeigt das Volumen biologischer Avocados ein starkes Wachstum, was auf eine steigende Nachfrage nach biologischen Produkten hinweist.
- Rückgang im Jahr 2018:
  - Beide Kategorien (konventionell und biologisch) zeigen einen deutlichen Rückgang im Jahr 2018. Dieser Rückgang könnte durch eine Vielzahl von Faktoren verursacht worden sein, wie z.B. ungünstige Wetterbedingungen, reduzierte Produktion, Handelshemmnisse oder eine Abnahme der Nachfrage.
- Verhältnis zwischen konventionellen und biologischen Avocados:
  - Obwohl das Volumen biologischer Avocados wächst, bleibt das Volumen konventioneller Avocados deutlich höher. Dies zeigt, dass der Großteil der Verbraucher weiterhin konventionelle Avocados bevorzugt.

### 4.4 Umsatz pro Region (Revenue per Region):

Der Umsatz wird berechnet, indem der `AveragePrice` mit dem `Total Volume` multipliziert wird.

In [45]:
# Erstellung einer Kopie des DataFrames
avocado_data = avocado_data.copy()

# Berechnung des Umsatzes für jede Zeile
avocado_data['Revenue'] = avocado_data['AveragePrice'] * avocado_data['Total Volume']
avocado_data.head()

,Date,AveragePrice,Total Volume,4046,4225,4770,Total Bags,Small Bags,Large Bags,XLarge Bags,type,year,region,Revenue
0,2015-12-27,1.33,64236.62,1036.74,54454.85,48.16,8696.87,8603.62,93.25,0.0,conventional,2015.0,Albany,85434.7046
1,2015-12-20,1.35,54876.98,674.28,44638.81,58.33,9505.56,9408.07,97.49,0.0,conventional,2015.0,Albany,74083.9230
2,2015-12-13,0.93,118220.22,794.70,109149.67,130.50,8145.35,8042.21,103.14,0.0,conventional,2015.0,Albany,109944.8046
3,2015-12-06,1.08,78992.15,1132.00,71976.41,72.58,5811.16,5677.40,133.76,0.0,conventional,2015.0,Albany,85311.5220
4,2015-11-29,1.28,51039.60,941.48,43838.39,75.78,6183.95,5986.26,197.69,0.0,conventional,2015.0,Albany,65330.6880


Nun können wir den Gesamtumsatz pro Region berechnen, indem wir die Revenue-Spalte nach der region-Spalte gruppieren und summieren:

In [48]:
# Gruppierung nach Region und Berechnung des Gesamtumsatzes
revenue_per_region = avocado_data.groupby('region')['Revenue'].sum().reset_index()

# Anzeige der ersten Einträge der Gruppierung
print("Umsatz pro Region:")
revenue_per_region.head()


Umsatz pro Region:


,region,Revenue
0,Albany,2.176672e+07
1,Atlanta,9.379337e+07
2,BaltimoreWashington,1.799084e+08
3,Boise,1.534667e+07
4,Boston,1.265429e+08


In [52]:
# Absteigend sortieren
revenue_per_region.sort_values(by="Revenue", ascending=False).head(10)

,region,Revenue
51,TotalUS,6.387593e+09
6,California,1.121414e+09
52,West,1.066834e+09
29,Northeast,9.600079e+08
45,SouthCentral,8.740593e+08
46,Southeast,7.036306e+08
15,GreatLakes,6.886618e+08
25,Midsouth,6.157238e+08
22,LosAngeles,4.842276e+08
35,Plains,3.600366e+08


**Interpretation:**

- Marktgröße und Verteilung:
  - `TotalUS` gibt den Gesamtrahmen des nationalen Marktes vor. Regionen wie `California`, `West` und `Northeast` tragen erheblich zu diesem Gesamtumsatz bei. Diese Regionen haben eine starke Nachfrage und sind für den Avocado-Markt von entscheidender Bedeutung.
- Hoher Umsatz in Kalifornien:
  - Kalifornien, als Hauptproduzent von Avocados, zeigt einen sehr hohen Umsatz. Dies spiegelt sowohl die Produktion als auch den Konsum innerhalb des Bundesstaates wider.
- Starke Nachfrage im Nordosten und Südosten:
  - Regionen wie `Northeast` und `Southeast` haben ebenfalls starke Märkte, was auf eine wachsende Beliebtheit von Avocados in städtischen und dicht besiedelten Gebieten hinweist.
- Wachstumschancen in kleineren Märkten:
  - Regionen wie `Plains` und `Midsouth` haben kleinere Marktanteile, könnten aber potenziell weiter wachsen, wenn die Nachfrage dort steigt oder gezielte Marketingmaßnahmen umgesetzt werden.

### 4.5 Preis-Volumen-Korrelation (Price-Volume Correlation):

Diese Korrelation hilft uns zu verstehen, ob es einen Zusammenhang zwischen dem Preis von Avocados und der Menge der verkauften Avocados gibt. Eine negative Korrelation könnte darauf hinweisen, dass höhere Preise zu geringeren Verkaufsvolumen führen, während eine positive Korrelation das Gegenteil bedeutet.

Der Korrelationskoeffizient kann zwischen -1 und 1 liegen:

- +1: Perfekte positive Korrelation (wenn der Preis steigt, steigt auch das Volumen).
- -1: Perfekte negative Korrelation (wenn der Preis steigt, sinkt das Volumen).
- 0: Keine Korrelation (kein Zusammenhang zwischen Preis und Volumen).

In [54]:
# Berechnung der Korrelation zwischen AveragePrice und Total Volume
price_volume_correlation = avocado_data['AveragePrice'].corr(avocado_data['Total Volume'])

print("Korrelation zwischen Durchschnittspreis und Gesamtvolumen:")
price_volume_correlation


Korrelation zwischen Durchschnittspreis und Gesamtvolumen:


np.float64(-0.19275238715271914)

**Interpretation:**

- Da der Wert negativ ist, deutet dies darauf hin, dass tendenziell ein leichter negativer Zusammenhang zwischen Preis und Volumen besteht. Das bedeutet:
  - Wenn der Preis steigt, nimmt das Verkaufsvolumen leicht ab
  - Wenn der Preis sinkt, nimmt das Verkaufsvolumen leicht zu.
- Schwache Korrelation (nahe 0):
  - Der Wert liegt nahe bei 0, was auf eine schwache Korrelation hindeutet. Das bedeutet, dass der Zusammenhang zwischen Preis und Volumen nicht stark ist.
  - In praktischen Worten bedeutet dies, dass der Preis nur einen geringen Einfluss auf die verkaufte Menge hat. Andere Faktoren wie Saisonalität, Verfügbarkeit, Nachfrage und Marketing könnten ebenfalls eine bedeutende Rolle spielen.

### 4.6 Saisonale Nachfrage (Seasonal Demand):

Man möchte herausfinden, ob es bestimmte Jahreszeiten oder Monate gibt, in denen die Nachfrage nach Avocados höher oder niedriger ist.

Zuerst erstellen wir neue Spalten für Monate und Wochen aus der Date-Spalte, um die saisonale Nachfrage besser zu analysieren:

In [55]:
# Neue Spalten für Monat und Woche erstellen
avocado_data['Month'] = avocado_data['Date'].dt.month
avocado_data['Week'] = avocado_data['Date'].dt.isocalendar().week

# Anzeige der ersten Zeilen zur Überprüfung
avocado_data[['Date', 'Month', 'Week', 'Total Volume', 'AveragePrice']].head()

,Date,Month,Week,Total Volume,AveragePrice
0,2015-12-27,12,52,64236.62,1.33
1,2015-12-20,12,51,54876.98,1.35
2,2015-12-13,12,50,118220.22,0.93
3,2015-12-06,12,49,78992.15,1.08
4,2015-11-29,11,48,51039.60,1.28


Wir berechnen das durchschnittliche Verkaufsvolumen und den durchschnittlichen Preis pro Monat:

In [56]:
# Gruppieren nach Monat und Berechnen des durchschnittlichen Volumens und Preises
monthly_data = avocado_data.groupby('Month').agg({
    'Total Volume': 'mean',  # Durchschnittliches Verkaufsvolumen pro Monat
    'AveragePrice': 'mean'   # Durchschnittlicher Preis pro Monat
}).reset_index()

# Anzeige der monatlichen aggregierten Daten
print("Durchschnittliches Verkaufsvolumen und Preis pro Monat:")
monthly_data


Durchschnittliches Verkaufsvolumen und Preis pro Monat:


,Month,Total Volume,AveragePrice
0,1,9.035654e+05,1.311019
1,2,1.018825e+06,1.274387
2,3,8.845054e+05,1.332255
3,4,8.801935e+05,1.374380
4,5,9.727150e+05,1.346601
5,6,9.293397e+05,1.408964
6,7,8.590482e+05,1.462077
7,8,8.036857e+05,1.512514
8,9,7.533192e+05,1.572654
9,10,6.803941e+05,1.579683


In [57]:
# Absteigend sortieren nach Preis
monthly_data.sort_values(by="AveragePrice", ascending=False)

,Month,Total Volume,AveragePrice
9,10,6.803941e+05,1.579683
8,9,7.533192e+05,1.572654
7,8,8.036857e+05,1.512514
6,7,8.590482e+05,1.462077
10,11,6.796073e+05,1.455563
5,6,9.293397e+05,1.408964
3,4,8.801935e+05,1.374380
4,5,9.727150e+05,1.346601
2,3,8.845054e+05,1.332255
11,12,7.746305e+05,1.331112


In [58]:
# Absteigend sortieren nach Volumen
monthly_data.sort_values(by="Total Volume", ascending=False)

,Month,Total Volume,AveragePrice
1,2,1.018825e+06,1.274387
4,5,9.727150e+05,1.346601
5,6,9.293397e+05,1.408964
0,1,9.035654e+05,1.311019
2,3,8.845054e+05,1.332255
3,4,8.801935e+05,1.374380
6,7,8.590482e+05,1.462077
7,8,8.036857e+05,1.512514
11,12,7.746305e+05,1.331112
8,9,7.533192e+05,1.572654


**Interpretation:**

- Saisonale Schwankungen im Volumen:
  - Die höchsten Verkaufsvolumina treten im Februar und Mai auf, was auf besondere Ereignisse oder Feiertage hinweist (z.B. Super Bowl, Cinco de Mayo).
  - Die niedrigsten Verkaufsvolumina sind in den Herbstmonaten (September und Oktober), was auf eine geringere Nachfrage oder ein geringeres Angebot hinweisen könnte.
- Saisonale Schwankungen im Preis:
  - Saisonale Schwankungen im Preis:
  - Die Preise sinken in den Wintermonaten (Dezember bis Februar), was möglicherweise durch eine höhere Verfügbarkeit oder ein geringeres Konsuminteresse verursacht wird.
- Höhere Preise trotz niedrigerem Volumen:
  - In Monaten wie August und September, in denen das Volumen niedrig ist, sind die Preise hoch. Dies deutet auf eine Angebotsverknappung hin.

### Zusatz: Jährliche Wachstumsrate (Annual Growth Rate):

In [53]:
# Schritt 1: Gruppieren nach Jahr und Berechnen des jährlichen Durchschnittspreises und Gesamtvolumens
annual_data = avocado_data.groupby('year').agg({
    'AveragePrice': 'mean',   # Durchschnittlicher Preis pro Jahr
    'Total Volume': 'sum'     # Gesamtvolumen pro Jahr
}).reset_index()

# Initialisieren der neuen Spalten für die Wachstumsraten
annual_data['Price Growth Rate'] = 0.0
annual_data['Volume Growth Rate'] = 0.0

# Schritt 2: Berechnung der Wachstumsraten für jedes Jahr, ab dem zweiten Jahr
for i in range(1, len(annual_data)):
    # Berechnung der Wachstumsrate des Durchschnittspreises
    previous_price = annual_data.loc[i - 1, 'AveragePrice']
    current_price = annual_data.loc[i, 'AveragePrice']
    if previous_price != 0:  # Vermeidung von Division durch 0
        annual_data.loc[i, 'Price Growth Rate'] = ((current_price - previous_price) / previous_price) * 100

    # Berechnung der Wachstumsrate des Verkaufsvolumens
    previous_volume = annual_data.loc[i - 1, 'Total Volume']
    current_volume = annual_data.loc[i, 'Total Volume']
    if previous_volume != 0:  # Vermeidung von Division durch 0
        annual_data.loc[i, 'Volume Growth Rate'] = ((current_volume - previous_volume) / previous_volume) * 100

# Schritt 3: Ausgabe der berechneten Wachstumsraten
print("Jährliche Wachstumsrate für Durchschnittspreis und Verkaufsvolumen:")
annual_data


Jährliche Wachstumsrate für Durchschnittspreis und Verkaufsvolumen:


,year,AveragePrice,Total Volume,Price Growth Rate,Volume Growth Rate
0,2015.0,1.375590,4.385469e+09,0.000000,0.000000
1,2016.0,1.338640,4.820890e+09,-2.686176,9.928727
2,2017.0,1.515128,4.934306e+09,13.184129,2.352591
3,2018.0,1.347531,1.382738e+09,-11.061558,-71.977043


## Schlussfolgerung und Empfehlung

**1. Saisonale Nachfrage (Seasonal Demand):**

- Ergebnisse:
  - Die höchsten Verkaufsvolumina treten im Februar und Mai auf, was auf spezielle Ereignisse wie den Super Bowl und Cinco de Mayo hindeutet. Diese Zeiträume sind für den Avocado-Verkauf besonders profitabel.
  - In den Herbstmonaten (September und Oktober) ist das Verkaufsvolumen am niedrigsten, während die Preise steigen. Dies könnte auf ein reduziertes Angebot oder eine geringere Nachfrage hindeuten.
- Empfehlung:
  - Strategische Planung: Erhöhen Sie die Lagerbestände und planen Sie gezielte Marketingkampagnen rund um wichtige Ereignisse wie den Super Bowl und Cinco de Mayo.
  - Preisstrategien: Während der Herbstmonate könnten Preisanpassungen oder Promotionsaktionen sinnvoll sein, um den Absatz zu steigern und hohe Preise auszugleichen.

**2. Saisonale Nachfrage (Seasonal Demand):**

- Ergebnisse:
  - Es besteht eine schwache negative Korrelation zwischen Preis und Volumen (-0,1927). Das bedeutet, dass der Preis nur einen geringen Einfluss auf die verkaufte Menge hat.
- Empfehlung:
  - Preisanpassung: Da die Korrelation schwach ist, könnte es sinnvoll sein, die Preise gezielt zu erhöhen, ohne einen starken Rückgang des Verkaufsvolumens zu befürchten.
  - Weitere Untersuchungen: Es wäre nützlich, spezifische Regionen oder Zeiträume zu analysieren, in denen diese Korrelation stärker ausgeprägt ist.

**3. Umsatz pro Region:**

- Ergebnisse:
  - Die Regionen mit dem höchsten Umsatz sind TotalUS, California und West. Diese Regionen haben ein hohes Marktpotenzial und sollten weiterhin im Fokus stehen.
- Empfehlung:
  - Regionale Expansion: Nutzen Sie das Potenzial der Regionen mit hohem Umsatz und fokussieren Sie sich auf Marketing und Vertrieb in diesen Gebieten.
  - Analyse von Nischenmärkten: Untersuchen Sie Regionen mit niedrigerem Umsatz, um mögliche Wachstumschancen zu identifizieren.

**4.  Verkaufsvolumen nach Typ (Volume by Type):**

- Ergebnisse:
  - Konventionelle Avocados haben ein deutlich höheres Verkaufsvolumen als biologische, aber das Wachstum im Bio-Segment ist signifikant.
- Empfehlung:
  - Förderung von Bio-Produkten: Erhöhen Sie das Angebot und die Sichtbarkeit von biologischen Avocados, um auf die wachsende Nachfrage nach nachhaltigen und biologischen Produkten zu reagieren.

